Blit with shader to avoid breaking render passes

zakorgy opened the pull request:
This depends on https://github.com/szeged/webrender/pull/337.
Also made some minor changes to not call `begin_render_pass` inside `clear_target_rects`

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/szeged/webrender/343)
<!-- Reviewable:end -->

kvark commented on webrender/src/device/gfx/device.rs:
```diff
@@ -2471,6 +2465,9 @@ impl<B: hal::Backend> Device<B> {
         filter: TextureFilter,
     ) {
         debug_assert!(self.inside_frame);
+        if self.inside_render_pass || self.bound_draw_fbo == DEFAULT_DRAW_FBO {
```
can we be rendering to a default FBO without an open render pass here? It looks like we should have only the first part as a condition, and the second would go into an `assert!` or something like that

zakorgy replied:
Thanks for noticing this, I think we don't need the second condition at all. The purpose of this PR is to have blits to non main targets as well, and blitting to `DEFAULT_DRAW_FBO` (main target) should always happen inside a render pass as well.

kvark commented on webrender/src/renderer.rs:
```diff
@@ -3905,13 +3911,19 @@ impl<B: hal::Backend> Renderer<B> {
                 )
             });
 
+            // Unfortunatelly this clear rect must executed in a separate render pass
```
let's reconsider this. What parts of the setup are different?

zakorgy replied:
Finaly managed to move this clear inside the next render pass in https://github.com/szeged/webrender/pull/343/commits/6eecba39c7ac073340f97ee8f3da138905f9647c

kvark commented on webrender/src/renderer.rs:
```diff
@@ -4516,7 +4545,17 @@ impl<B: hal::Backend> Renderer<B> {
-        self.device.begin_render_pass(transit_to_present && target.alpha_batch_containers.is_empty());
+        {
+            if target_kind == TargetKind::Main || !target.empty_without_batches() {
+                self.device.begin_render_pass(transit_to_present && target.alpha_batch_containers.is_empty());
```
this reads a bit misplaced to me: by this moment we should have already done the clears and blits, both of which require the render pass. So why don't we make sure that the render pass starts earlier and here assume it's already set up?

zakorgy replied:
The clear above clears the entire texture (because the clear rect is none with gfx backends: https://github.com/szeged/webrender/pull/343/files#diff-f5062b694b9fe53fc1757ed483d577d9R4528) so we will use a clear load operation in this case, and we record that load operation in `clear_target` prior to render pass begin. So that clear must be outside of render pass. Because that clear is outside of a render pass, The blit is outside of a render pass as well in this case we don't force to blot with a shader and use the default blit call.

zakorgy commented:
@kvark I've also fixed that disturbing `begin_render_pass` logic as we discussed yesterday: collected the clears, removed the first render pass begin inside the if statement and execute them inside the second render pass:
https://github.com/szeged/webrender/pull/343/commits/578c81b64ba1c3a1d695b5f3885767c16e58d6e9#diff-f5062b694b9fe53fc1757ed483d577d9R4330-R4396
